### DISCUSSION TITLE
Intel AMT not available when using PostgreSQL?

### DISCUSSION THREAD

**Discussion by Salvoxia** [Ylianst/MeshCentral#6321] [Issues] [answered]
Hi everyone,

TL;DR:  MeshCentral does not seem to start the AMT Management Node when configured with PostgreSQL. Is this a bug and if not, which databases can be used in conjunction with AMT?

I've been using MeshCentral with Agent-enabled clients for quite some time now. My setup is as follows:
  - MeshCentral in Docker 
  - Behind Traefik as reverse Proxy
  - PostgreSQL as database
  - LDAP for authentication

Recently I tried adding a new AMT enabled machine as AMT-only agent-less type. After enabling AMT (ACM) on the machine itself it was picked by MeshCentral using the "Scan Network" function, it was added, but showed as offline with no information whatsoever.
![image](https://github.com/user-attachments/assets/a7b2f89b-9b1b-49fe-995c-978743ce70ef)
![image](https://github.com/user-attachments/assets/e41fbd73-71ad-4849-a423-7b88ccba5784)

I've spent the last two days troubleshooting to find out what's going on.
Facts:
  - AMT Webpage of machine is reachable on port 16993
  - MeshCommander Desktop installation works as expected
  - Another bare-metal MeshCentral installation with minimal configuration works as expected

After gradually adjusting the config of the bare-metal installation to match my original Docker-based installation more and more I found the point where AMT stopped working: When switching from the embedded SQLite3 database to PostgreSQL.

Runnning MeshCentral with `--debug amt` I noticed that the following line shows up when using SQLite3, but __does not__ show up when using PostgreSQL with everything else identical:
```
AMT: Start Management node//qjmXWx8bhOxs6BEvs@2ZxmGKsT8j57PuibqDY5Doqj$M3nTf8XTo4aKTACQNDB4C 3
```
If that line showed up everything worked, if it didn't the agent-less AMT machines were "dead" inside MeshCentral.

Now my question: Is that expected and if so, which databases can be used in conjcunation AMT and which cannot?

Best Regards,
Salvoxia

**Comment by si458** [comment]
You should be able to use ANY database?
This sounds like possibly a bug or something mis-configured.
Alot of people are reporting using agent only groups don't seem to work properly and I can't figure out why?
What amt device do u have?
What version is it?
How did u activate the amt?
Can u share ur config.json? (Hide secrets)
Can u share ur reverseproxy settings/config? 
Is the amt port directly exposed to the Internet or is it proxied via ur reverse proxy?
Are u in lan or wan or hybrid mode for meshcentral?

**Reply by Salvoxia** [reply]
Thank you for your swift response, I'll work with you as best as I can!

> You should be able to use ANY database?

That's what I assumed as well, it got me pretty confused when I narrowed it down to just the database being different..

> What amt device do u have?

I am using an Intel/ASUS NUC 13 Pro Kit NUC13L3Kv5 (Core i5-1350P).

> What version is it?
How did u activate the amt?

The ME Firmware version reports as 16.1.30-build 2361 and I enabled it manually in the machine's UEFI (Redirections for SoL, KVM and Storage enabled, no user consent, Network enabled, static IP address, IP and host name matching what the OS is set up with).

> Can u share ur config.json? (Hide secrets)

My config with PostgreSQL and AMT __not__ working:
```json
{
  "$schema": "https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json",
  "settings": {
    "cert": "meshcentral-test.mydomain.internal",
    "plugins":{"enabled": false},
    "LANonly": true,
    "sessionKey": "xxxxxxxxxxxxxx",
    "TLSOffload": false,
    "SelfUpdate": false,
    "AllowFraming": false,
    "WebRTC": false,
    "meshErrorLogPath": "/opt/meshcentral/meshcentral-data/logs",
    "postgres": {
      "host": "meshcentral-test-db.mydomain.internal",
      "user": "meshcentral",
      "password": "meshcentral",
      "port": 5556,
      "database": "meshcentral"
    },
    "autoBackup": {
      "backupIntervalHours": false
    }
  },
  "domains": {
    "": {
      "NewAccounts": true,
      "userNameIsEmail": true
    }
  }
}
```
This config produces the following output:
```bash
~$ node node_modules/meshcentral --debug amt
MeshCentral HTTP redirection server running on port 80.
MeshCentral v1.1.27, LAN mode.
MeshCentral HTTPS server running on port 443.
```

And the same config with `postgres` 'commented out' as `_postgres` and AMT working:
```json
{
  "$schema": "https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json",
  "settings": {
    "cert": "meshcentral-test.mydomain.internal",
    "plugins":{"enabled": false},
    "LANonly": true,
    "sessionKey": "xxxxxxxxxxxxxx",
    "TLSOffload": false,
    "SelfUpdate": false,
    "AllowFraming": false,
    "WebRTC": false,
    "meshErrorLogPath": "/opt/meshcentral/meshcentral-data/logs",
    "_postgres": {
      "host": "meshcentral-test-db.mydomain.internal",
      "user": "meshcentral",
      "password": "meshcentral",
      "port": 5556,
      "database": "meshcentral"
    },
    "autoBackup": {
      "backupIntervalHours": false
    }
  },
  "domains": {
    "": {
      "NewAccounts": true,
      "userNameIsEmail": true
    }
  }
}
```

This config produces the following output:
```bash
~$ node node_modules/meshcentral --debug amt
MeshCentral HTTP redirection server running on port 80.
MeshCentral v1.1.27, LAN mode.
MeshCentral HTTPS server running on port 443.
AMT: Start Management node//Q@v@B1UiRLVf6Ooljtu7qon7o8PYkIb8@WNceoEdpvDwvoBdIPt9UaxczB$A$GfS 3
AMT: nuc13.mydomain.internal Checking Intel AMT state...
AMT: nuc13.mydomain.internal Attempt Initial Contact Local
AMT: nuc13.mydomain.internal Attempt Initial Local Contact 3 nuc13.mydomain.internal
AMT: nuc13.mydomain.internal Direct-Connect TLS nuc13.mydomain.internal admin
AMT: nuc13.mydomain.internal Initial Contact Response 200
AMT: nuc13.mydomain.internal Intel AMT connected with TLS.
AMT: nuc13.mydomain.internal Fetching hardware inventory.
AMT: nuc13.mydomain.internal Done.
```

> Can u share ur reverseproxy settings/config?
Is the amt port directly exposed to the Internet or is it proxied via ur reverse proxy?

In the minimal config above I eliminated LDAP and the reverse proxy, so this should be out of the equation.

> Are u in lan or wan or hybrid mode for meshcentral?

LAN mode as per the minimal config above.

**Reply by si458** [reply]
Hmmm...
1. what node version are u using?
2. I think ur first config.json showing not working is wrong/bad copy+paste as its showing `_postgres`
3. I'm sure u should be seeing an extra line saying the AMT is enabled and running on port 4433 which isn't shown here?
4. U should really remove the `tlsoffload: false` as u should be using ssl! Even if it's self-signed that's OK!

**Reply by Salvoxia** [reply]
> what node version are u using?

```bash
:~$ node --version
v18.19.0
```
But keep in mind the official Docker image shows the same behavior, so whatever node version comes with that..

> I think ur first config.json showing not working is wrong/bad copy+paste as its showing _postgres

You're right, that should be `postgres` there. Fixed it in my comment above.

> I'm sure u should be seeing an extra line saying the AMT is enabled and running on port 4433 which isn't shown here?

Nope, nothing more than what I posted. The [User Guide](https://meshcentral.com/docs/MeshCentral2UserGuide.pdf) States that MPS is disabled in LAN mode, which is what I'm running. And if I understand that right, MPS is only important for provisioning or CIRA? I need neither.

> U should really remove the tlsoffload: false as u should be using ssl! Even if it's self-signed that's OK!

Well, this is just a dummy test installation to get to the bottom of this issue, in my real installation TLS is termianted at the reverse proxy and no direct ports to the Meshcentral container are open.  
Anyway, I removed the line from my test config and repeated the test, same output. So does not seem to be relevant for this issue,

**Reply by si458** [reply]
Sorry, I'm still learning all the amt things while maintaining meshcentral!
Yes I do think ur right, the port 4433 is only for the cira or wan mode etc not lan mode only!
When u add the device, in the general tab, does it show as activated acm? With maybe no credentials or invalid credentials next to it?
What happens if u remove the lanonly, and let ir run in hybrid mode?
Dows the device connect at all?
It's hard to try and replicate the issue when u don't have hardware to test :(
I only have an amt 7, 8, 12 device, nothing newer!
Just waiting for someone to kindly donate me a device so I can fully test what's going on! (Happy to pay import tax if I have too)

**Reply by Salvoxia** [reply]
> Sorry, I'm still learning all the amt things while maintaining meshcentral!

No worries, this has been my first trip with AMT as well. And I'm very grateful for your work on the project!

> When u add the device, in the general tab, does it show as activated acm? With maybe no credentials or invalid credentials next to it?

Adding the device like so:
![image](https://github.com/user-attachments/assets/fdb02350-6c60-4114-93de-a41ad0a1e2de)
General Tab looks like this:
![image](https://github.com/user-attachments/assets/a33814bc-cffc-4282-b553-32481be970fc)
It seems there's no communication with the device happening at all.

> What happens if u remove the lanonly, and let ir run in hybrid mode?
Dows the device connect at all?

Now MPS server is started
```bash
:~$ node node_modules/meshcentral --debug amt
MeshCentral HTTP redirection server running on port 80.
MeshCentral v1.1.27, Hybrid (LAN + WAN) mode.
MeshCentral Intel(R) AMT server running on meshcentral-test.mydomain.internal:4433.
MeshCentral HTTPS server running on meshcentral-test.mydomain.internal:443.
```
However, the AMT device still looks the same in General Tab, Details tab also has nothing
![image](https://github.com/user-attachments/assets/812ca704-c898-482c-aecb-3a71130185e6)

> It's hard to try and replicate the issue when u don't have hardware to test :(
I only have an amt 7, 8, 12 device, nothing newer!

Do you think this has something to do with the AMT version? I also used Wireshark to capture the communication between MeshCentral and the AMT device in both cases - With PostgreSQL it seemed MeshCentral didn't even try to reach out to the device.

> Just waiting for someone to kindly donate me a device so I can fully test what's going on! (Happy to pay import tax if I have too)

I understand Intel made some significant changes with the removal of SMB mode. Do you know that "oldest" (read: "cheapest") device with a sufficiently up-to-date ME firmware version is that you could use for debugging/dev?

**Reply by si458** [reply]
I will try have a look tomorrow if I get chance with my AMT 12 device.
Random throught,
can ur meshcentral device ping the dns name u set for your amt device?
Can u also access the amt port of ur remote device remotely from meshcentral or ur Web browser?
`curl -v https://remoteamtdevice.dnsname.internal:16993`

**Reply by Salvoxia** [reply]
> I will try have a look tomorrow if I get chance with my AMT 12 device.

Great, thanks!

>Random throught,
can ur meshcentral device ping the dns name u set for your amt device?

```bash
~$ ping nuc13.mydomain.internal
PING nuc13.mydomain.internal (192.168.1.50) 56(84) bytes of data.
64 bytes from nuc13.mydomain.internal (192.168.1.50): icmp_seq=1 ttl=64 time=1.55 ms
```

> Can u also access the amt port of ur remote device remotely from meshcentral or ur Web browser?
curl -v https://remoteamtdevice.dnsname.internal:16993

Had to add `-k` because it's still using the default Intel AMT certificate:
```bash
~$ curl -v -k https://nuc13.mydomain.internal:16993
*   Trying 192.168.1.50:16993...
* Connected to nuc13.mydomain.internal (192.168.1.50) port 16993 (#0)
* ALPN: offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN: server did not agree on a protocol. Uses default.
* Server certificate:
*  subject: C=US; ST=California; L=Santa Clara; O=Intel Corporation; CN=Intel▒ AMT self-signed certificate
*  start date: Jan  1 00:00:00 2022 GMT
*  expire date: Dec 31 23:59:59 2037 GMT
*  issuer: C=US; ST=California; L=Santa Clara; O=Intel Corporation; CN=Intel▒ AMT self-signed certificate
*  SSL certificate verify result: self-signed certificate (18), continuing anyway.
* using HTTP/1.x
> GET / HTTP/1.1
> Host: nuc13.mydomain.internal:16993
> User-Agent: curl/7.88.1
> Accept: */*
>
< HTTP/1.1 303 See Other
< Location: /logon.htm
< Content-Length: 0
< Server: Intel(R) Active Management Technology 16.1.30.2361
<
* Connection #0 to host nuc13.mydomain.internal left intact
```

Forget about everything I wrote about Docker and Reverse Proxy, these are all eliminated now.
It's the same bare-metal installation with just two configs, without `postgres` it works, with it doesn't. So the network prerequisites are there.

If we can find some reasonably priced device that suits your needs, I might be inclined to be that donator if it helps you keeping AMT working with MeshCentral.

**Reply by si458** [reply]
The issue could be because it's activated to ur old setup, which uses the local db, but then u switched to it to postgres, and it's a new setup as such!

So u might need to deactivate amt (you can use the meshcmd for this so long as u know the amt user/pass)

Also one thing u can try is, delete the postgres database,
Set it bk to using nedb,
then do a db export
https://ylianst.github.io/MeshCentral/meshcentral/#database-export
Then set postgres in config.json
Then import the database
https://ylianst.github.io/MeshCentral/meshcentral/#database-import

**Reply by Salvoxia** [reply]
I'm open to trying this, but there's a couple of items why I doubt this will change anything:
  - During testing i repeatedly wiped (deleted) the local SQLite3 database with which it was  working, each time it after the DB was created cleanly it worked again after adding the device
  - At the same time it works with a local MeshCommander installation as well as a [mesh-mini](https://github.com/BrytonSalisbury/mesh-mini) installation (literally simultaneously)
 
I don't think there's some kind of "pairing" going on if no provisioning server or CIRA is used...

Nevertheless, I'll test that tomorrow! Thanks for your suggestions!

**Reply by Salvoxia** [reply]
I just performed the database migration you suggested by exporting the SQLite3 DB and importing it into a fresh PostgreSQL DB.
Now MeshCentral shows the last known information about the AMT device (e.g. version), but does not update it. The debug output remains the same.
```bash
~$ node node_modules/meshcentral --debug amt
MeshCentral HTTP redirection server running on port 80.
MeshCentral v1.1.27, Hybrid (LAN + WAN) mode.
MeshCentral Intel(R) AMT server running on meshcentral-test.mydomain.internal:4433.
MeshCentral HTTPS server running on meshcentral-test.mydomain.internal:443.
```

**Reply by si458** [reply]
weirdly enough ive just spotted a weird bug with the AMT here on my version 12,
when i add my device into the web ui using the scan network, and then give it the password,
it takes about 2 mins to show infomation and change from trying credentials to activated
but this is because AMT 13 or above now uses TLS where as 12 or below uses NO-TLS,
and the default for local mode is always use TLS to begin with!
so i have to sit n wait before it actually adds the device in or shows up!

p.d. this is using ANY database tho, not just 1 certain DB, but still looking into things slowly

**Reply by si458** [reply]
ok so ignore my previous post,
the is an extra option in config.json to force meshcentral to always use no-tls to speed up the process,
this wont help with newer devices that are amt 16+ tho sadly
`
Starting from Alder Lake platforms with Raptor Lake CPUs running Intel CSME 16.1 firmware, remote connections to Intel AMT unsecure TCP/IP ports 16992, 16994 and 623 are no longer supported. TLS ports 16993, 16995 and 664 must be used for connecting to Intel AMT. Starting from Intel CSME 19 firmware on Arrow Lake platforms, connecting to Intel AMT without TLS is not supported at all (i.e., also local connections must use TLS ports).
`
```
 "domains": {
    "": {
      "_title": "MyServer",
      "_title2": "Servername",
      "_minify": true,
      "_newAccounts": true,
      "_userNameIsEmail": true,
      "amtManager": {
        "TlsConnections": false
      }
    }
  },
```

**Reply by Salvoxia** [reply]
I always added the AMT device manually using `Add Local` and explicitly selecting `TLS security required`, so it should have tried to connect to the device using TLS from the beginning.

**Reply by si458** [reply]
ok how very weird !?
so after my testing locally with nedb,
then switched to postgres and starting a fresh,
it does indeed not attempt to connect to the remote computer!?
WTF is going on how weird!?

**Reply by Salvoxia** [reply]
I'm very glad you were able to reproduce it, that's the very first step!

**Reply by si458** [reply]
ive found the problem, but im confused as if the codes right (and not changed at all)
local amt has **never** worked for mysql/mariadb, postgres and sqlite3 for 3 years !?
the reason it works for most is because the build-in default DB is NeDB,
and the old guides pdfs all tell people to use mongodb for peering or distribution

**Reply by Salvoxia** [reply]
I'm sorry I might have added to the confusion by thinking the default database was SQLite3. 
But I guess this shows that the amount of people who
  - use one of the mentioned databases
  - use local AMT
  - care enough to report something not working
is quite low.

My medium-term plans are to move towards peering as well, but I plan on having a backup installation for which AMT is the most important feature.

**Reply by si458** [reply]
yep just confirmed with mysql and sqlite3, doesnt work there either, so nice catch!
as for peering, to peer meshcentral servers you need to use mongodb
https://ylianst.github.io/MeshCentral/meshcentral/#server-peering

**Comment by si458** [comment]
fix has been pushed for you! https://github.com/Ylianst/MeshCentral/commit/77f44fc308c41647b911453aa8bda62b68c6f1e5
its actually an sql query! it was always looking for devices which had uuids for intel amt,
HOWEVER you never know the uuid until you connect/first connect!

**Reply by Salvoxia** [reply]
Thanks for that fix!
I cloned the repo and built a local Docker image to test with.
Now it peridiocally logs the following error
```
SQLERR5 TypeError: Cannot read properties of undefined (reading 'tls')
    at /opt/meshcentral/meshcentral/amtscanner.js:185:74
    at /opt/meshcentral/meshcentral/db.js:2094:268
    at Query.<anonymous> (/opt/meshcentral/meshcentral/db.js:1354:39)
    at Query.handleReadyForQuery (/opt/meshcentral/meshcentral/node_modules/pg/lib/query.js:138:12)
    at Client._handleReadyForQuery (/opt/meshcentral/meshcentral/node_modules/pg/lib/client.js:290:19)
    at Connection.emit (node:events:519:28)
    at /opt/meshcentral/meshcentral/node_modules/pg/lib/connection.js:114:12
    at Parser.parse (/opt/meshcentral/meshcentral/node_modules/pg-protocol/dist/parser.js:36:17)
    at Socket.<anonymous> (/opt/meshcentral/meshcentral/node_modules/pg-protocol/dist/index.js:11:42)
    at Socket.emit (node:events:519:28)
```

```/opt/meshcentral/meshcentral/db.js:2094:268``` seems to point to a piece of code that was changed in that commit.
Any ideas?

**Reply by si458** [reply]
hmm it seems to think one of the hosts it found in the db is not an AMT device!
do you have a mix of devices?
ones with amt, ones without amt, ones with just meshagent, ones with meshagent+amt, etc?

**Reply by Salvoxia** [reply]
I have two device groups
  - Managed using a software agent
  - Agentless AMT only
In the first group I have a couple of devices and I tried adding my AMT device to the second one.

Even after deleting the Agentless AMT only group that error is logged every 30s

**Reply by si458** [reply, marked as answer]
fixed! https://github.com/Ylianst/MeshCentral/commit/adeaac1588224849da4f2278fd743f56f053f24a
my bad! it was returning all devices and not checking if they actually have the intelamt flag!
try pulling/rebuilding or wait 10mins and the `master` docker image should be auto build :)

**Reply by Salvoxia** [reply]
It's working! That's great! Thanks you so much for effort!
I guess you got yourself a new sponsor :)

**Reply by si458** [reply]
glad its sorted and got there in the end!
am just amazed nobody else noticed or mentioned this bug in past few years!
but i will check all other issues tomorrow tho in my free time!
P.S: donation/sponsor links are all here hehe https://www.si458.co.uk/2024/01/05/donation/ ❤️ 
P.SS: the amtscanner runs every 30 seconds, and checks for new machines and state changes etc.
hense the error every 30 seconds(ish)

**Reply by si458** [reply]
@Salvoxia i would just like to say **THANK YOU SO MUCH!**
I just received your donation today! it really did make a tear run down my eye ❤️ 
totally unexpected!
i wasnt sure how to contact you, so put the message here and on telegram

**Reply by Salvoxia** [reply]
Thank YOU for spending your free time to maintain this great project!
